### PR TITLE
Global styles: add onChange actions to color palette items

### LINKED ISSUE #44138: Global Styles: Can't open Colour Picker by clicking on colour circle
### Description

In Global Styles one can't open the Colour Picker by clicking on the colour circle. You need to first click on the ellipsis and then press Edit which is confusing since the colour circles are interactive but do nothing. It's okay for editing the name to be buried but changing the colour should be quicker.

### Step-by-step reproduction instructions

1. Go to Site Editor → Global Styles
2. Click _Colors_
3. Open a pallette
4. Click on one of the colour circles

Expected result: The colour picker should open.
Actual result: Nothing happens.

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/612155/190030765-29cedc2d-25e8-4e32-b0ec-e9f7dd2349fa.mp4

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ramonjd**
## What?
Resolves https://github.com/WordPress/gutenberg/issues/44138

The PR opens the color picker when palette item is selected.

Renames the "Edit colors/gradients" label with "Show details" (suggestions for alternatives welcome).

# How?

By inserting a new instance of the color picker popover.

We track the currently-edited color/gradient by its index in the colors/gradients collection. The collection is rendered to the page according to this index. 

We need the index of the currently-edited color/gradient to ensure we're editing the correct item in the collection. 

We can't check by value or slug because two or more items might share these values. This is why this PR also updates the color/gradient picker component to return the index of the item to the onChange callback. Using this value we can lookup the currently-edited item in the colors/gradients collection.

As a fallback, if index doesn't match an item OR the matched item's value does not equal the newly-chosen value, we just open the list view as before.

## Why?
Before nothing happened when you clicked on a color or gradient option. Weird!

## Testing Instructions

1. Go to Site Editor → Global Styles
2. Click Colors
3. Open a palette
4. Click on one of the colour circles

The color picker should show for the color you clicked on, and you should be able to adjust the color successfully. 

Test it out with gradients too!

## Screenshots or screencast 

https://user-images.githubusercontent.com/6458278/202356776-bd339a94-9eee-461d-9ab9-1e1ff748cd25.mp4

### CODE REVIEW

**Comment by scruffian** [comment]
If I can edit the theme colors when the list is closed, then what is the purpose of having the ability to "edit" the colors anymore?

**Comment by ramonjd** [comment]
> If I can edit the theme colors when the list is closed, then what is the purpose of having the ability to "edit" the colors anymore?

Good question, something I didn't think of.

This is probably the advantage of @pbking's approach in https://github.com/WordPress/gutenberg/pull/45657 - removing it completely. 

I guess there's only the question of whether we need the "Edit" functionality for either PR 😄

**Comment by ramonjd** [comment]
Another idea to improve this PR from @scruffian and @SaxonF was to change the label of "Edit" to "List view" indicate that there's an alternative view (given that this PR allows direct editing of colors)

**Comment by scruffian** [comment]
The benefit of the "edit" mode is that it shows a list of the items with names, so it is useful to be able to switch between states, but the label should be "List view" or similar, not edit.

@SaxonF

**Comment by scruffian** [comment]
A couple of observations:

1. When I click to edit, the view is switched from the circles to the list view. Is there a way to edit the colors without switching to the list view?
2. Is there a better name than "list view". We already use that name for... the list view!

**Comment by ramonjd** [comment]
> When I click to edit, the view is switched from the circles to the list view. Is there a way to edit the colors without switching to the list view?

Good idea. This would be better, I agree. 

I've done some hacking and it looks possible, but it requires a bit of a refactor and will bloat this PR a bit. 

I can take it as far as I can and we can assess.


> Is there a better name than "list view". We already use that name for... the list view!

Hmm... If we allow edit in place then the list view becomes redundant, aside from showing the palette item names.

Maybe "Show color/gradient labels?" That's a bit wordy. 

"Show more", "Show details", "Where is my thesaurus?"

File: packages/components/src/palette-edit/index.js [hunk 1]
```diff
@@ -313,6 +313,13 @@ export default function PaletteEdit( {
 		! elements[ editingElement ].slug;
 	const elementsLength = elements.length;
 	const hasElements = elementsLength > 0;
+	const editPaletteItem = ( value, collection ) => {
+		const valueIndex = collection.findIndex(
+			( item ) => item.color === value
```
Comment (ramonjd) [inline packages/components/src/palette-edit/index.js]: Problem here: there could be two or more colors/gradients in the presets array with the same color/gradient value.

Also we should be checking for `item.gradient` for gradients, but that's beside the point 😄

**Comment by ramonjd** [comment]
> When I click to edit, the view is switched from the circles to the list view. Is there a way to edit the colors without switching to the list view?
> Is there a better name than "list view". We already use that name for... the list view!

I've taken a stab at this in 581ea0b87bda3208cd5050f56d9ef2e27fa534ac

It's messy, but it seems to work 😄

**Comment by noisysocks** [comment]
It works! Nice job.

It definitely feels nicer now that _something_ happens when you click on a colour circle in Palette UI. It felt really broken before.

I do wonder if this makes changing an existing colour _too_ easy. It's not something I expect users to do often. (I could be wrong?) It also greatly increases the likelihood that a colour is named incorrectly. For example, it's now really easy to set the "Pale pink" colour to be blue without renaming "Pale pink" to "Pale blue" or whatever.

Part of me wonders if we should just disable the button altogether so that it's a completely non-interactive colour circle until you enter edit mode.

What do you think @jasmussen?

**Comment by ramonjd** [comment]
Thanks for taking a look at this @noisysocks 

> I do wonder if this makes changing an existing colour too easy. It's not something I expect users to do often. (I could be wrong?) It also greatly increases the likelihood that a colour is named incorrectly. For example, it's now really easy to set the "Pale pink" colour to be blue without renaming "Pale pink" to "Pale blue" or whatever.

Good point. This would apply only to custom palettes though, right? That is, colors a user has already actively added. Maybe it's not as invasive (?)

Also, the commit history of this PR (see [comment](https://github.com/WordPress/gutenberg/pull/45681#issuecomment-1316352226)) has a version that opens the edit "list view" when clicking on a color, so maybe there's a compromise there.


> Part of me wonders if we should just disable the button altogether so that it's a completely non-interactive colour circle until you enter edit mode.

Another valid option 😄 It's the apparent "clickability" that's causing confusion for me personally, so making it obvious that the palette options aren't to be clicked might also work.

**Comment by noisysocks** [comment]
> Good point. This would apply only to custom palettes though, right? That is, colors a user has already actively added. Maybe it's not as invasive (?)

When I try this PR out locally you can edit all palettes not just the custom one. If we limited it to the custom palette then that would definitely alleviate my concern.

**Comment by ramonjd** [comment]
> When I try this PR out locally you can edit all palettes not just the custom one. If we limited it to the custom palette then that would definitely alleviate my concern.

Sorry, my bad. My brain was thinking "changing the text label" and not the color. Withdraw my comment 😄

**Comment by jasmussen** [comment]
> It definitely feels nicer now that something happens when you click on a colour circle in Palette UI. It felt really broken before.

I would tend to agree, right now it feels really weird that I can select a spot color but then, nothing. As a small improvement, this seems like a step in the right direction.

But I do wonder: if the only remaining thing that the edit flow allows, is to rename colors, why not just show that? It's not a strong opinion, and by still gating it, it might make for a simpler experience in 80% of cases. CC: @mtias

**Comment by ramonjd** [comment]
>  if the only remaining thing that the edit flow allows, is to rename colors, why not just show that? It's not a strong opinion, and by still gating it, it might make for a simpler experience in 80% of cases.

I think renaming colors only applies for custom colors (?), so theme and default colors in edit mode allow editing colors.

The simplicity of https://github.com/WordPress/gutenberg/pull/45657, which shows the editable list view all the time, appeals to me personally.

I think the only concern raised what that it has the potential to  take up a fair thwack of real estate, and, as a consequence,  remove the opportunity to add more items to the sidebar in that specific view.

As an example, here what I'm seeing in 2023 with a few custom colors:



The browser viewport height is 1100px

**Review by scruffian — APPROVED** [review]
I think this is a good improvement. 🚢

**Comment by noisysocks** [comment]
Yeah, way better than what's in `trunk`. Let's get this in and go from there.

**Comment by ciampo** [comment]
👋  Hey there!

Would it be possible to bolster unit tests to `PaletteEdit`, especially after the changes from this PR? (we could add the missing CHANGELOG entry from this PR too!

**Comment by ramonjd** [comment]
> Would it be possible to bolster unit tests to PaletteEdit, especially after the changes from this PR? (we could add the missing CHANGELOG entry from this PR too!

I always forget the changelog! Thanks for the reminder 👍
